File: C#-OOP-Basics/Problems Course Instance 2018/Inheritance/Mankind/Worker.cs
using System;
using System.Text;

public class Worker : Human
{
    private decimal weekSalary;
    private decimal workHours;

    public Worker(string firstName, string lastName, decimal weekSalary, decimal workHoursPerDay)
        : base(firstName, lastName)
    {
        this.WeekSalary = weekSalary;
        this.WorkHours = workHoursPerDay;
    }

    public decimal WeekSalary
    {
        get
        {
            return this.weekSalary;
        }

        private set
        {
            if (value <= 10)
            {
                throw new ArgumentException("Expected value mismatch! Argument: weekSalary");
            }

            this.weekSalary = value;
        }
    }

    public decimal WorkHours
    {
        get
        {
            return this.workHours;
        }

        private set
        {
            if (value < 1 || value > 12)
            {
                throw new ArgumentException("Expected value mismatch! Argument: workHoursPerDay");
            }

            this.workHours = value;
        }
    }

    public override string ToString()
    {
        StringBuilder sb = new StringBuilder();
        sb.Append(base.ToString());
        sb.AppendLine($"Week Salary: {this.weekSalary:F2}");
        sb.AppendLine($"Hours per day: {this.workHours:F2}");
        sb.AppendLine($"Salary per hour: {this.CalculateSalary():F2}");

        return sb.ToString();
    }

    private decimal CalculateSalary()
    {
        return this.weekSalary / (5m * this.workHours);
    }
}

File: Java-OOP-Basics/Problems Course Instance 2016_2017/src/Lesson03Inheritance/Exercise/pr03_mankind/Worker.java
package Lesson03Inheritance.Exercise.pr03_mankind;

public class Worker extends Human {

    private static final String TOO_SHORT_LAST_NAME = "Expected length more than 3 symbols!Argument: lastName";
    private static final String INVALID_WEEK_SALARY = "Expected value mismatch!Argument: weekSalary";
    private static final String INVALID_WORK_HOURS_PER_DAY = "Expected value mismatch!Argument: workHoursPerDay";
    private static final double MIN_WEEK_SALARY = 10d;
    private static final int MAX_WORK_HOURS = 12;
    private static final int MIN_WORK_HOURS = 1;

    private double weekSalary;
    private double workHoursPerDay;

    public Worker(String firstName, String lastName, double weekSalary, double workHoursPerDay) {
        super(firstName, lastName);
        this.setWeekSalary(weekSalary);
        this.setWorkHoursPerDay(workHoursPerDay);
    }

    private void setWeekSalary(double weekSalary) {
        if (weekSalary < MIN_WEEK_SALARY) {
            throw new IllegalArgumentException(INVALID_WEEK_SALARY);
        }
        this.weekSalary = weekSalary;
    }

    private void setWorkHoursPerDay(double workHoursPerDay) {
        if (workHoursPerDay < MIN_WORK_HOURS || workHoursPerDay > MAX_WORK_HOURS) {
            throw new IllegalArgumentException(INVALID_WORK_HOURS_PER_DAY);
        }
        this.workHoursPerDay = workHoursPerDay;
    }

    private double getSalaryPerHour() {
        return this.weekSalary / this.workHoursPerDay / 7d;
    }

    @Override
    protected void setLastName(String lastName) {
        if (lastName == null || lastName.trim().length() < 4) {
            throw new IllegalArgumentException(TOO_SHORT_LAST_NAME);
        }
        super.setLastName(lastName);
    }

    @Override
    protected String getInfo() {
        return String.format("%s%nSalary:  %.2f%nHours per day: %.2f%nSalary per hour: %.2f",
                super.getInfo(), this.weekSalary, this.workHoursPerDay, this.getSalaryPerHour());
    }
}
